### DISCUSSION TITLE
Cannot inject PSR-7 ServerRequestInterface in Middleware

### DISCUSSION THREAD

**Discussion by georgms** [laravel/framework#40043] [Ideas]
- Laravel Version: 8.76.1
- PHP Version: 8.0.13
- Database Driver & Version: Irrelevant

### Description:

It's possible to inject a [PSR-7 `ServerRequestInterface` into controller methods](https://laravel.com/docs/8.x/requests#psr7-requests). However, the same does not work with Middlewares. The `handle` method will always receive an `Illuminate\Http\Request` despite this signature:

```php
use Psr\Http\Message\ServerRequestInterface;

…

public function handle(ServerRequestInterface $request, Closure $next): ResponseInterface
{
  …
}
```

### Steps To Reproduce:

1. Create a Middleware with the signature `public function handle(ServerRequestInterface $request, Closure $next): ResponseInterface`.
2. Enable the Middleware in the Routing / Kernel.
3. Fire a request that will trigger the middleware.

#### Expected result

The `handle` method receives a `ServerRequestInterface`.

#### Actual result

```
TypeError: App\Http\Middleware\MyMiddleware::handle(): Argument #1 ($request) must be of type Psr\Http\Message\ServerRequestInterface, Illuminate\Http\Request given, called in […]/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php on line 167 and defined in […]/app/Http/Middleware/MyMiddleware.php:60
```

**Comment by henzeb** [comment]
Don't know why this popped up on the first page, but to answer the question: the middleware is part of a pipeline. In this case the pipeline is receiving an instance of `Illuminate\Http\Request`. There is no dependency injection going on in there.  Read more about pipelines here: https://dev.to/abrardev99/pipeline-pattern-in-laravel-278p
